test/Test/ChainFragment.hs

```diff
@@ -0,0 +1,737 @@
+      -- removing the block gives the original
+      && case CF.headPoint c of
+           Nothing -> True
+           Just headPnt -> CF.rollback headPnt c' == Just c
+      && CF.drop 1 c' == c
+      -- chain is one longer
+      && CF.length c' == CF.length c + 1
+
+prop_rollback :: TestChainFragmentAndPoint -> Bool
+prop_rollback (TestChainFragmentAndPoint c p) =
+    case CF.rollback p c of
+      Nothing -> True
+      Just c' ->
+        -- chain is a prefix of original
+           CF.isPrefixOf c' c
+        -- chain head point is the rollback point
+        && CF.headPoint c' == Just p
+
+prop_rollback_head :: TestBlockChainFragment -> Bool
+prop_rollback_head (TestBlockChainFragment c) =
+  case CF.headPoint c of
+    Nothing      -> True
+    Just headPnt -> CF.rollback headPnt c == Just c
+
+prop_successorBlock :: TestChainFragmentAndPoint -> Property
+prop_successorBlock (TestChainFragmentAndPoint c p) =
+  CF.pointOnChainFragment p c ==>
+  case CF.successorBlock p c of
+    Nothing -> CF.headPoint c === Just p
+    Just b  -> property $ CF.pointOnChainFragment (CF.blockPoint b) c
+
+prop_lookupBySlot :: TestChainFragmentAndPoint -> Property
+prop_lookupBySlot (TestChainFragmentAndPoint c p) =
+  CF.pointOnChainFragment p c ==>
+    property $ case CF.lookupBySlot c (pointSlot p) of
+      Just b  -> CF.pointOnChainFragment (CF.blockPoint b) c
+              && CF.blockSlot b == pointSlot p
+      Nothing -> False
+
+prop_intersectChainFragments :: TestChainFragmentFork -> Bool
+prop_intersectChainFragments (TestChainFragmentFork origL1 origL2 c1 c2) =
+  case CF.intersectChainFragments c1 c2 of
+    Nothing -> (CF.null origL1 || CF.null origL2)
+            && L.intersect (CF.toNewestFirst c1) (CF.toNewestFirst c2) == []
+    Just (l1, l2, r1, r2) ->
+      l1 == origL1 && l2 == origL2 &&
+      CF.headPoint l1 == CF.headPoint l2 &&
+      CF.toNewestFirst r1 ++ CF.toNewestFirst l1 == CF.toNewestFirst c1 &&
+      CF.toNewestFirst r2 ++ CF.toNewestFirst l2 == CF.toNewestFirst c2
+
+
+prop_serialise_chain :: TestBlockChainFragment -> Property
+prop_serialise_chain (TestBlockChainFragment chain) =
+  prop_serialise chain
+
+prop_pointOnChainFragment :: TestChainFragmentAndPoint -> Bool
+prop_pointOnChainFragment (TestChainFragmentAndPoint c p) =
+  CF.pointOnChainFragment p c == CF.pointOnChainFragmentSpec p c
+
+prop_splitBeforeSlot_length :: TestChainFragmentAndPoint -> Property
+prop_splitBeforeSlot_length (TestChainFragmentAndPoint c p) =
+  case CF.splitBeforeSlot c (pointSlot p) of
+    Just (l, r) -> CF.length l + CF.length r === CF.length c
+    Nothing     -> property $ not (CF.pointOnChainFragment p c)
+
+prop_splitBeforeSlot_pointOnChainFragment :: TestChainFragmentAndPoint -> Bool
+prop_splitBeforeSlot_pointOnChainFragment (TestChainFragmentAndPoint c p) =
+  case CF.splitBeforeSlot c (pointSlot p) of
+    Just (l, r) -> not (CF.pointOnChainFragment p l)
+                && CF.joinChainFragments l r == Just c
```

Comment:
Still, even though it's partial, the join should result in a `Just c`. I have applied your suggestion about comparing the joins instead of the sum of the lengths.